name.abuchen.portfolio/src/name/abuchen/portfolio/online/impl/variableurl/Factory.java

```diff
@@ -1,54 +1,103 @@
 package name.abuchen.portfolio.online.impl.variableurl;
 
+import name.abuchen.portfolio.online.impl.variableurl.macros.ConstString;
+import name.abuchen.portfolio.online.impl.variableurl.macros.Currency;
+import name.abuchen.portfolio.online.impl.variableurl.macros.FormattedDate;
+import name.abuchen.portfolio.online.impl.variableurl.macros.ISIN;
+import name.abuchen.portfolio.online.impl.variableurl.macros.Macro;
+import name.abuchen.portfolio.online.impl.variableurl.macros.PageNumber;
+import name.abuchen.portfolio.online.impl.variableurl.macros.TickerSymbol;
+import name.abuchen.portfolio.online.impl.variableurl.macros.WKN;
+import name.abuchen.portfolio.online.impl.variableurl.urls.ConstURL;
+import name.abuchen.portfolio.online.impl.variableurl.urls.VariableURL;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
 import java.util.LinkedList;
 import java.util.List;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
 public class Factory
 {
+    private interface MacroParser
+    {
+        Macro parse(CharSequence input) throws IllegalArgumentException;
+    }
+
     private static final Pattern URL_MACRO_ABSTRACT = Pattern.compile("\\{(.*?)}"); //$NON-NLS-1$
+    private static final Collection<MacroParser> MACRO_PARSERS = Arrays.asList(
+        FormattedDate::new,
+        PageNumber::new,
+        ISIN::new,
+        WKN::new,
+        TickerSymbol::new,
+        Currency::new
+    );
 
     private Factory()
     {}
 
     public static VariableURL fromString(CharSequence input)
     {
-        List<CharSequence> parts = new LinkedList<>();
         Matcher abstractMatcher = URL_MACRO_ABSTRACT.matcher(input);
         int lastEnd = 0;
+        CharSequence part;
+        List<Macro> macros = new LinkedList<>();
+        Macro macro = null;
 
         while (abstractMatcher.find())
         {
-            parts.add(input.subSequence(lastEnd, abstractMatcher.start()));
-            lastEnd = abstractMatcher.end();
+            part = input.subSequence(lastEnd, abstractMatcher.start());
 
-            parts.add(abstractMatcher.group(1));
-        }
+            if (part.length() > 0)
+                macros.add(new ConstString(part));
 
-        parts.add(input.subSequence(lastEnd, input.length()));
+            lastEnd = abstractMatcher.end();
+            part = abstractMatcher.group(1);
 
-        if (parts.size() > 1)
-        {
-            try
+            for (MacroParser macroParser : MACRO_PARSERS)
             {
-                return new FormattedDate(parts);
-            }
-            catch (IllegalArgumentException ignored)
-            {
-                // try next pattern
-            }
+                try
+                {
+                    macro = macroParser.parse(part);
+                }
+                catch (IllegalArgumentException e)
```

Comment:
Ich habe die Change gemergt. Eigentlich bin ich ja kein Fan von "control flow via exceptions". Exceptions sind relativ teuer anzulegen. In diesem Fall macht es vermutlich keinen sonderlich grossen Unterschied - so häufig wird der Code nicht durchlaufen. Hier hätte ich eher ein ```marcoParser.isApplicable(part)``` oder ähnliches angedacht.

Wie auch immer - die Changes sind drin. Wird das Leben einfacher machen! Danke!